django/db/models/sql/compiler.py

```diff
@@ -287,7 +289,11 @@ def get_select_from_parent(klass_info):
             get_select_from_parent(klass_info)
 
         ret = []
+        col_idx = 1
         for col, alias in select:
+            if alias is None and with_col_aliases:
+                alias = f"col{col_idx}"
+                col_idx += 1
```

Comment:
It's unused for empty results, so maybe we can move it below the `try...except` block:
```python
            try:
                sql, params = self.compile(col)
            except EmptyResultSet:
                empty_result_set_value = getattr(
                    col, "empty_result_set_value", NotImplemented
                )
                if empty_result_set_value is NotImplemented:
                    # Select a predicate that's always False.
                    sql, params = "0", ()
                else:
                    sql, params = self.compile(Value(empty_result_set_value))
            else:
                sql, params = col.select_format(self, sql, params)
            if alias is None and with_col_aliases:
                alias = f"col{col_idx}"
                col_idx += 1
            ret.append((col, (sql, params), alias))
```

Reply:
I'm not sure I understand, what exactly is unused here? Not against moving it after the SQL generation logic but an alias will be required in all cases when `with_col_aliases` is passed.

Reply:
Ahh, true. I've convinced myself that `alias` is unused if `EmptyResultSet` is raised :facepalm:

Reply:
👍 I don't mind moving it after in all cases, I like that it matches how a select component reads where the SQL precedes its alias.

Reply:
Moved in #15947.